packages/core/src/common/message-rpc/msg-pack-extension-manager.ts

```diff
@@ -0,0 +1,87 @@
+// *****************************************************************************
+// Copyright (C) 2022 STMicroelectronics and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { addExtension } from 'msgpackr';
+import { ResponseError } from './rpc-message-encoder';
+
+/**
+ * Handles the global registration of custom MsgPackR extensions
+ * required for the default RPC communication. MsgPackR extensions
+ * are installed globally on both ends of the communication channel.
+ * (frontend-backend, pluginExt-pluginMain).
+ * Is implemented as singleton as it is  also used in plugin child processes which have no access to inversify.
+ */
+export class MsgPackExtensionManager {
+    private static readonly INSTANCE = new MsgPackExtensionManager();
+    public static getInstance(): MsgPackExtensionManager {
+        return this.INSTANCE;
+    }
```

Comment:
>First I think singleton instances/APIs should be declared outside of the class describing the singleton: Where we can use Inversify then the singleton instance can be constructed there. For runtimes that aren't using Inversify, then a singleton instance can be constructed/defined specifically for them.

Sure we can do that. However, I just wanted to notice that the  current architecture of the core Theia messaging proxies doesn't really rely on inversify. Everything starting from the ProxyFactory, over to the RPCProtocol and the underlying channels is directly constructed. So even though we have access to inversify we probably end up with constructing a singleton instance.

>Second, with the way things are done here, the extension for ResponseError will only be registered if this script happens to be loaded/imported somehow. As I don't see anyone importing this script yet, I'm quite curious to understand how this can be working?

Good catch, its probably just workig because its indirectly imported in the index.ts of the plugin-ext package. I will have a look at this.

>Overall I don't think we should rely on the global packer/unpacker but rather have components depending on those to have it injected, with or without Inversify. Customization should then be done before injection.

In fact we don't rely on the global packr/unpackr and [already construct](https://github.com/eclipse-theia/theia/blob/d5d67db470a88738fdac9593fd6e0393a206b2c8/packages/core/src/common/message-rpc/rpc-message-encoder.ts#L127) dedicated packr/unpackr isntances. However the extension mechanism in msgpackR still only allows global registration. So you cannot configure extensions per packr/unpackR instance at the moment.

Reply:
@paul-marechal Are you fine with the way this extension mechanism is currently implemented (even though it is not ideal) or do you have another solution in mind?